mammoth/modules/adapters.py

```diff
@@ -0,0 +1,227 @@
+"""
+Inject small bottleneck layers with residual connection.
+Can be applied during main training,
+or injected into an already trained network to adapt it for a new task.
+"""
+
+import torch
+import torch.nn as nn
+from collections import defaultdict
+from typing import Union, Set, Dict
+from functools import partial
+
+from x_transformers.x_transformers import (
+    AttentionLayers,
+    FeedForward,
+    Residual,
+    SimpleRMSNorm,
+)
+
+
+class FeedForwardAdapterLayer(nn.Module):
+    """A separate adapter layer injected after a FeedForward. Has its own norms."""
+    def __init__(self, dim, pre_norm=True, sandwich_norm=False, layer_dropout=0.0, **kwargs):
+        super().__init__()
+        norm_fn = partial(SimpleRMSNorm, dim)
+        self.pre_branch_norm = norm_fn() if pre_norm else None
+        self.post_branch_norm = norm_fn() if sandwich_norm else None
+        self.post_main_norm = norm_fn() if not pre_norm else None
+        self.ff = FeedForward(dim, **kwargs)
+        self.residual = Residual(dim, scale_residual=False)
+        self.layer_dropout = layer_dropout
+
+    @property
+    def is_wrapper(self):
+        return False
+
+    def as_layer_struct(self):
+        return nn.ModuleList([
+            nn.ModuleList([
+                self.pre_branch_norm,
+                self.post_branch_norm,
+                self.post_main_norm,
+            ]),
+            self.ff,
+            self.residual,
+        ])
+
+    def apply(self, tmp_layer_types, tmp_layer_structs, tmp_layer_dropouts):
+        # FeedForwards are injected after the base ff
+        tmp_layer_types.append('f')
+        tmp_layer_structs.append(self.as_layer_struct())
+        tmp_layer_dropouts.append(self.layer_dropout)
+        return tmp_layer_types, tmp_layer_structs, tmp_layer_dropouts
+
+
+class LoraAdapterLayer(nn.Module):
```

Comment:
we didn't support lora previously, did we? if so, might be good to explicitly mention in the global description of the PR